src/AppInstallerCLICore/Workflows/WorkflowBase.cpp

```diff
@@ -444,6 +444,42 @@ namespace AppInstaller::CLI::Workflow
         }
     }
 
+    void OpenSourceFor::operator()(Execution::Context& context) const
+    {
+        switch (m_operationType)
+        {
+        case OperationType::Search:
+        case OperationType::Show:
+            context << 
+                Workflow::OpenSource();
+            break;
+
+        case OperationType::Install:
+            context << Workflow::OpenSource();
+            if (!context.Args.Contains(Execution::Args::Type::Force))
+            {
+                context << Workflow::OpenCompositeSource(Repository::PredefinedSource::Installed, false, Repository::CompositeSearchBehavior::AvailablePackages);
+            }
+            break;
+        case OperationType::List:
+        case OperationType::Uninstall:
+        case OperationType::Upgrade:
+            context <<
+                Workflow::OpenSource() <<
+                Workflow::OpenCompositeSource(Workflow::DetermineInstalledSource(context));
+            break;
+        case OperationType::Pin:
+        case OperationType::Export:
+        case OperationType::Completion:

```

Comment:
I'm not sure about `Completion`. There are other cases like search where we don't use the installed, and I see you had to use Search type there

Reply:
Hmm. I can understand that point. One thing that I don't understand is the difference between the `PredefinedSource::Installed` and `DetermineInstalledSource`. The cases where completion uses the Predefined source are for the commands which use `DetermineInstalledSource`. 

Perhaps what would make the most sense is to have completion throw the invalid state and keep the current behavior of explicit management of sources for completion

Reply:
DetermineInstalledSource will filter the installed source to include only user scope | machine scope | default (previous behavior where for win32, user and machine scopes are included, for msix, only apps registered to user are included)